Allow the usage of lambdas for `InputPort` default values (#3465)
aiidateam/aiida-core: docs/source/working/processes.rst

.. code:: python

    spec.input('positive_number', required=False, default=lambda: Int(1), valid_type=(Int, Float), validator=is_number_positive)

> Here we define an input named ``positive_number`` that is not required, if a value is not explicitly passed, the default will be used and if a value *is* passed, it should be of type ``Int`` or ``Float`` and it should be valid according to the ``is_number_positive`` validator.

Reviewer: ```suggestion
Here we define an input named ``positive_number`` that should be of type ``Int`` or ``Float`` and should pass the test of the ``is_number_positive`` validator.
If no value is passed, the default will be used.
```

.. warning::

>     In python, it is important to `never use mutable defaults for function arguments <https://docs.python.org/3/reference/compound_stmts.html#function-definitions>`_, since they are instantiated at function definition and reused for each invocation.

Reviewer: ```suggestion
    In python, it is good practise to avoid mutable defaults for function arguments, `since they are instantiated at function definition and reused for each invocation. <https://docs.python.org/3/reference/compound_stmts.html#function-definitions>`_
```

>     This can lead to unexpected results where the default value can change between function calls.

Reviewer: ```suggestion
    This can lead to unexpected results when the default value is changed between function calls.
```

>     The same concept holds for process ports in AiiDA and so defaults should *only use immutable types*.

Reviewer: ```suggestion
    In the context of AiiDA, nodes (both stored and unstored) are considered *mutable* and should therefore *not* be used as default values for process ports.
```

>     A node, whether unstored or stored, is considered mutable and so should not be used as a default value.

Reviewer: ```suggestion
```
